test/dtls_mtu_test.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <openssl/dtls1.h>
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+
+/* for SSL_USE_ETM() */
+#include "../ssl/ssl_locl.h"
+
+static int debug = 0;
+
+static unsigned int clnt_psk_callback(SSL *ssl, const char *hint,
+				      char *ident, unsigned int max_ident_len,
+				      unsigned char *psk,
+				      unsigned int max_psk_len)
+{
+	snprintf(ident, max_ident_len, "psk");
+
+	if (max_psk_len > 20)
+		max_psk_len = 20;
+	memset(psk, 0x5a, max_psk_len);
+
+	return max_psk_len;
+}
+
+static unsigned int srvr_psk_callback(SSL *ssl, const char *identity,
+				      unsigned char *psk,
+				      unsigned int max_psk_len)
+{
+	if (max_psk_len > 20)
+		max_psk_len = 20;
+	memset(psk, 0x5a, max_psk_len);
+	return max_psk_len;
+}
+
+static int mtu_test(SSL_CTX *ctx, const char *cs, int no_etm)
+{
+	SSL *srvr_ssl = NULL, *clnt_ssl = NULL;
+	BIO *sc_bio = NULL, *cs_bio = NULL;
+	int ret;
+	int i, j;
+	int mtus[30];
+	unsigned char buf[530];
+	int rv = 0;
+
+	memset(buf, 0x5a, sizeof(buf));
+
+	sc_bio = BIO_new(BIO_s_mem());
+	cs_bio = BIO_new(BIO_s_mem());
+
+	/* Non-blocking */
+	BIO_set_mem_eof_return(sc_bio, -1);
+	BIO_set_mem_eof_return(cs_bio, -1);
+
+	clnt_ssl = SSL_new(ctx);
+	srvr_ssl = SSL_new(ctx);
+
+	if (no_etm)
+		SSL_set_options(srvr_ssl, SSL_OP_NO_ENCRYPT_THEN_MAC);
+
+	if (SSL_set_cipher_list(srvr_ssl, cs) != 1 ||
+	    SSL_set_cipher_list(clnt_ssl, cs) != 1) {
+		ERR_print_errors_fp(stdout);
+		goto out;
+	}
+
+
+	/* The BIOs very freed twice, once from each SSL. Compensate. */
+	BIO_up_ref(sc_bio);
+	BIO_up_ref(cs_bio);
+
+	SSL_set_bio(clnt_ssl, cs_bio, sc_bio);
+	SSL_set_bio(srvr_ssl, sc_bio, cs_bio);
+
+	while (1) {
+		/* Client first, then server. And client finishes last */
+		ret = SSL_connect(clnt_ssl);
+		if (ret == 1)
+			break;
+		else if (SSL_get_error(clnt_ssl, ret) != SSL_ERROR_WANT_READ) {
+			ERR_print_errors_fp(stdout);
+			goto out;
+		}
+
+		ret = SSL_accept(srvr_ssl);
+		if (ret == 1) {
+			if (debug)
+				printf("Server established connection\n");
+		} else if (SSL_get_error(srvr_ssl, ret) != SSL_ERROR_WANT_READ) {
+			ERR_print_errors_fp(stdout);
+			goto out;
+		}
+
+	}
+	if (debug)
+		printf("Channel established\n");
```

Comment:
I think all of the above boilerplate can be replaced with calls to some of the functions in ssltestlib.c, e.g. call create_ssl_ctx_pair(), create_ssl_objects() and create_ssl_connection(). See execute_test_large_message() in sslapittest.c for an example of usage.